microdnf-plugins/builddep_plugin/builddep.cpp

```diff
@@ -0,0 +1,263 @@
+            auto split = libdnf::utils::string::split(value, " ", 2);
+            if (split.size() != 2) {
+                throw libdnf::cli::ArgumentParserError(
+                    "Invalid value for macro definition \"{}\". \"macro expr\" format expected.", value);
+            }
+            rpm_macros.emplace_back(std::move(split[0]), std::move(split[1]));
+            return true;
+        });
+    cmd.register_named_arg(defs);
+}
+
+
+void BuildDepCommand::parse_builddep_specs(int specs_count, const char * const specs[]) {
+    const std::string_view ext_spec(".spec");
+    const std::string_view ext_srpm(".src.rpm");
+    const std::string_view ext_nosrpm(".nosrc.rpm");
+    std::set<std::string> unique_items;
+    for (int i = 0; i < specs_count; ++i) {
+        const std::string_view spec(specs[i]);
+        if (auto [it, inserted] = unique_items.emplace(spec); inserted) {
+            // TODO(mblaha): download remote URLs to temporary location + remove them afterwards
+            if (spec.ends_with(ext_spec)) {
+                spec_file_paths.emplace_back(spec);
+            } else if (spec.ends_with(ext_srpm) || spec.ends_with(ext_nosrpm)) {
+                srpm_file_paths.emplace_back(spec);
+            } else {
+                pkg_specs.emplace_back(spec);
+            }
+        }
+    }
+}
+
+
+bool BuildDepCommand::add_from_spec_file(std::set<std::string> & specs, const char * spec_file_name) {
+    auto spec = rpmSpecParse(spec_file_name, RPMSPEC_ANYARCH | RPMSPEC_FORCE, nullptr);
+    if (spec == nullptr) {
+        std::cout << "Failed to parse spec file \"" << spec_file_name << "\"." << std::endl;
+        return false;
+    }
+    auto dependency_set = rpmdsInit(rpmSpecDS(spec, RPMTAG_REQUIRENAME));
+    while (rpmdsNext(dependency_set) >= 0) {
+        specs.emplace(rpmdsDNEVR(dependency_set) + 2);
+    }
+    rpmdsFree(dependency_set);
+    rpmSpecFree(spec);
+    return true;
+}
+
+
+bool BuildDepCommand::add_from_srpm_file(std::set<std::string> & specs, const char * srpm_file_name) {
+    auto fd = Fopen(srpm_file_name, "r");
+    if (fd == NULL || Ferror(fd)) {
+        std::cerr << "Failed to open \"" << srpm_file_name << "\": " << Fstrerror(fd) << std::endl;
+        if (fd) {
+            Fclose(fd);
+            fd = nullptr;
+        }
+        return false;
+    }
+
+    Header header;
+    auto ts = rpmtsCreate();
+    rpmtsSetVSFlags(ts, _RPMVSF_NOSIGNATURES | _RPMVSF_NODIGESTS);
+    auto rc = rpmReadPackageFile(ts, fd, nullptr, &header);
+    rpmtsFree(ts);
+    Fclose(fd);
+    fd = nullptr;
+
+    if (rc == RPMRC_OK) {
+        auto dependency_set = rpmdsInit(rpmdsNewPool(nullptr, header, RPMTAG_REQUIRENAME, 0));
+        while (rpmdsNext(dependency_set) >= 0) {
+            std::string_view reldep = rpmdsDNEVR(dependency_set) + 2;
+            if (!reldep.starts_with("rpmlib(")) {
```

Comment:
The `string_view::starts_with` method has been supported since C++20.
I'm not sure if we no longer need C++17 compatibility.
We use our `libdnf::utils::string::starts_with` function in libdnf.

Reply:
After discussion, we came to the conclusion that we no longer need compatibility with C++17. We will use C++20.
So we can keep this code.